middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,57 @@
+import socket
+
+from mypy.types import Iterable
+
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_message_receive(connection: socket.socket, public_key: bytes) -> Iterable[AbstractMiddlemanMessage]:
+    """
+    Helper generator for receiving Middleman protocol messages.
+
+    The flow is:
+    1. Receive bytes until there is enough data to read payload length.
+    2. Receive more bytes until amount of bytes equal to payload_length is received.
+    3. Deserialize message and yield it.
+    4. Receive more bytes until separator is met.
+
+    If at any point separator is met, reset received data and restart.
+
+    Returns deserialized message payload.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(public_key, bytes)
+
+    received_data = b''
+    message_with_payload_length = None
+
+    while True:
+        received_data += connection.recv(1)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[:-FRAME_SEPARATOR_LENGTH] == FRAME_SEPARATOR
+        ):
+            received_data = b''
+            message_with_payload_length = None
+
+        # When 74 bytes are received, read payload_length. 71-74 are bytes with payload length.
+        if len(received_data) == 74:
+            payload_length = AbstractMiddlemanMessage.get_frame_format().signed_part_of_the_frame.payload_length.parse(
+                received_data[70:74]
+            )
+            message_with_payload_length = 74 + payload_length
```

Comment:
> But it is payload length, not frame.

My point was that `message_with_payload_length` sounds like the content of the message with the extra 4 bytes from the length field appended at the end. Not like the total length of the frame.

`frame_length` would indicate that it's the length and that it applies to the whole frame.

> `message > frame > payload`

Wait, what? For me it's `frame > payload = message`. Message is what is being sent as payload. In most cases it's a Golem message. Frame is what we use in the wire protocol to transport the actual message.

I think it's best to avoid using the word message from now on because it's confusing.

Reply:
> Wait, what? For me it's frame > payload = message. Message is what is being sent as payload. In most cases it's a Golem message. Frame is what we use in the wire protocol to transport the actual message.

That's why I have described how I am calling it inside code.

About `message_with_payload_length`, I will rename it to `header_plus_payload_length`.